File: clients/node-client/src/models/TaskResponse.ts
import { Chunkr } from "../Chunkr";
import { Configuration } from "./Configuration";
import { TaskResponseData, Output, Status } from "./TaskResponseData";

/**
 * Represents a response from a Chunkr API task operation.
 * Contains methods for polling, updating, and managing task state.
 */
export class TaskResponse implements TaskResponseData {
  #chunkr: Chunkr;

  public task_id!: string;
  public status!: Status;
  public created_at!: string;
  public finished_at!: string | null;
  public expires_at!: string | null;
  public message!: string;
  public input_file_url!: string | null;
  public pdf_url!: string | null;
  public output!: Output | null;
  public task_url!: string | null;
  public configuration!: Configuration;
  public file_name!: string | null;
  public page_count!: number | null;
  public error?: string;

  constructor(data: TaskResponseData, chunkr: Chunkr) {
    this.#chunkr = chunkr;
    Object.assign(this, data);
  }

  toJSON() {
    return {
      task_id: this.task_id,
      status: this.status,
      created_at: this.created_at,
      finished_at: this.finished_at,
      expires_at: this.expires_at,
      message: this.message,
      input_file_url: this.input_file_url,
      pdf_url: this.pdf_url,
      output: this.output,
      task_url: this.task_url,
      file_name: this.file_name,
      page_count: this.page_count,
      error: this.error,
    };
  }

  /**
   * Poll the task until it reaches a terminal state (Succeeded, Failed, or Cancelled).
   * @param {number} [interval=1000] - Polling interval in milliseconds
   * @returns {Promise<TaskResponse>} The completed task response
   * @throws {Error} If the task fails or reaches an unexpected state
   */
  async poll(interval: number = 1000): Promise<TaskResponse> {
    const pollingStates = [Status.STARTING, Status.PROCESSING];
    const terminalStates = [Status.SUCCEEDED, Status.FAILED, Status.CANCELLED];

    while (pollingStates.includes(this.status)) {
      try {
        await new Promise((resolve) => setTimeout(resolve, interval));
        const response = await this.#chunkr.getTask(this.task_id);
        Object.assign(this, response);
        console.debug(`Task ${this.task_id} status: ${this.status}`);
      } catch (error) {
        console.warn(`Polling error for task ${this.task_id}:`, error);
        continue;
      }
    }

    if (!terminalStates.includes(this.status)) {
      throw new Error(`Task ended in unexpected state: ${this.status}`);
    }

    if (this.status === Status.FAILED) {
      throw new Error(
        this.error || "Task failed without specific error message",
      );
    }

    return this;
  }

  /**
   * Cancel the current task. Only works if the task hasn't started processing.
   * @returns {Promise<void>}
   * @throws {Error} If the task has already started processing
   */
  async cancel(): Promise<void> {
    await this.#chunkr.cancelTask(this.task_id);
    await this.poll();
  }

  /**
   * Delete the current task.
   * @returns {Promise<void>}
   * @throws {Error} If the task is currently processing
   */
  async delete(): Promise<void> {
    await this.#chunkr.deleteTask(this.task_id);
  }

  /**
   * Get content from the task's output chunks.
   * @param {"html" | "markdown" | "content"} type - The type of content to retrieve
   * @returns {string} The concatenated content of all chunks
   */
  getContent(type: "html" | "markdown" | "content"): string {
    if (!this.output?.chunks) {
      return "";
    }

    return this.output.chunks
      .flatMap((chunk) =>
        chunk.segments.map((segment) => segment[type]).filter(Boolean),
      )
      .join("\n");
  }

  /**
   * Get HTML content from the task's output chunks.
   * @returns {string} The concatenated HTML content of all chunks
   */
  getHtml(): string {
    return this.getContent("html");
  }

  /**
   * Get Markdown content from the task's output chunks.
   * @returns {string} The concatenated Markdown content of all chunks
   */
  getMarkdown(): string {
    return this.getContent("markdown");
  }
}

File: clients/java-client/src/main/java/ai/chunkr/client/models/TaskResponse.java
package ai.chunkr.client.models;

import ai.chunkr.client.enums.Status;


import java.time.LocalDateTime;
import java.util.List;
import java.util.stream.Collectors;

/**
 * Represents a response from a Chunkr API task operation.
 * Contains methods for polling, updating, and managing task state.
 */
public class TaskResponse {
    
    private String taskId;
    
    private Status status;
    
    private Configuration configuration;
    
    private String createdAt;
    
    private String finishedAt;
    
    private String expiresAt;
    
    private String message;
    
    private Output output;
    
    private String taskUrl;
    
    private String error;

    private Object client; // ChunkrClient - using Object to avoid circular dependency

    // Constructors
    public TaskResponse() {}

    public TaskResponse(Object client) {
        this.client = client;
    }

    // Getters and Setters
    public String getTaskId() {
        return taskId;
    }

    public void setTaskId(String taskId) {
        this.taskId = taskId;
    }

    public Status getStatus() {
        return status;
    }

    public void setStatus(Status status) {
        this.status = status;
    }

    public Configuration getConfiguration() {
        return configuration;
    }

    public void setConfiguration(Configuration configuration) {
        this.configuration = configuration;
    }

    public String getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(String createdAt) {
        this.createdAt = createdAt;
    }

    public String getFinishedAt() {
        return finishedAt;
    }

    public void setFinishedAt(String finishedAt) {
        this.finishedAt = finishedAt;
    }

    public String getExpiresAt() {
        return expiresAt;
    }

    public void setExpiresAt(String expiresAt) {
        this.expiresAt = expiresAt;
    }

    public String getMessage() {
        return message;
    }

    public void setMessage(String message) {
        this.message = message;
    }

    public Output getOutput() {
        return output;
    }

    public void setOutput(Output output) {
        this.output = output;
    }

    public String getTaskUrl() {
        return taskUrl;
    }

    public void setTaskUrl(String taskUrl) {
        this.taskUrl = taskUrl;
    }

    public String getError() {
        return error;
    }

    public void setError(String error) {
        this.error = error;
    }

    public Object getClient() {
        return client;
    }

    public void setClient(Object client) {
        this.client = client;
    }

    /**
     * Poll the task until it reaches a terminal state (Succeeded, Failed, or Cancelled).
     * @param intervalMs Polling interval in milliseconds (default: 1000)
     * @return The completed task response
     * @throws RuntimeException If the task fails or reaches an unexpected state
     * @throws InterruptedException If the polling is interrupted
     */
    public TaskResponse poll(long intervalMs) throws InterruptedException {
        while (status.isProcessing()) {
            try {
                Thread.sleep(intervalMs);
                // TODO: Implement polling when ChunkrClient is available
                // TaskResponse updated = client.getTask(taskId);
                // Update all fields from the new response
                // this.status = updated.status;
                // this.finishedAt = updated.finishedAt;
                // this.message = updated.message;
                // this.output = updated.output;
                // this.error = updated.error;
                
                // For now, just break to avoid infinite loop  
                System.out.println("Task " + taskId + " status: " + status);
                break;
            } catch (Exception e) {
                System.err.println("Polling error for task " + taskId + ": " + e.getMessage());
                // Continue polling on error
            }
        }

        if (!status.isTerminal()) {
            throw new RuntimeException("Task ended in unexpected state: " + status);
        }

        if (status == Status.FAILED) {
            throw new RuntimeException(error != null ? error : "Task failed without specific error message");
        }

        return this;
    }

    /**
     * Poll the task with default interval of 1000ms.
     * @return The completed task response
     * @throws RuntimeException If the task fails or reaches an unexpected state
     * @throws InterruptedException If the polling is interrupted
     */
    public TaskResponse poll() throws InterruptedException {
        return poll(1000);
    }

    /**
     * Cancel the current task. Only works if the task hasn't started processing.
     * @throws RuntimeException If the task has already started processing
     */
    public void cancel() {
        try {
            // client.cancelTask(taskId); // TODO: Fix when ChunkrClient is available
            poll();
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw new RuntimeException("Task cancellation interrupted", e);
        }
    }

    /**
     * Delete the current task.
     * @throws RuntimeException If the task is currently processing
     */
    public void delete() {
        // client.deleteTask(taskId); // TODO: Fix when ChunkrClient is available
    }

    /**
     * Get content from the task's output chunks.
     * @param type The type of content to retrieve ("html", "markdown", or "content")
     * @return The concatenated content of all chunks
     */
    public String getContent(String type) {
        if (output == null || output.getChunks() == null) {
            return "";
        }

        return output.getChunks().stream()
                .flatMap(chunk -> chunk.getSegments().stream())
                .map(segment -> {
                    switch (type.toLowerCase()) {
                        case "html":
                            return segment.getHtml();
                        case "markdown":
                            return segment.getMarkdown();
                        case "content":
                        default:
                            return segment.getContent();
                    }
                })
                .filter(content -> content != null && !content.trim().isEmpty())
                .collect(Collectors.joining("\n"));
    }

    /**
     * Get HTML content from the task's output chunks.
     * @return The concatenated HTML content of all chunks
     */
    public String getHtml() {
        return getContent("html");
    }

    /**
     * Get Markdown content from the task's output chunks.
     * @return The concatenated Markdown content of all chunks
     */
    public String getMarkdown() {
        return getContent("markdown");
    }

    /**
     * Get raw content from the task's output chunks.
     * @return The concatenated raw content of all chunks
     */
    public String getContent() {
        return getContent("content");
    }

    @Override
    public String toString() {
        return "TaskResponse{" +
                "taskId='" + taskId + '\'' +
                ", status=" + status +
                ", createdAt='" + createdAt + '\'' +
                ", finishedAt='" + finishedAt + '\'' +
                ", message='" + message + '\'' +
                ", error='" + error + '\'' +
                '}';
    }
}